### PR TITLE
Add support for podGroup number limits for one queue

### LINKED ISSUE #452: [feature] Add support for podGroup number limits for one queue
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
> /kind feature
 /kind feature

**What happened**:
For scenarios of BigData and ML, the batch jobs often summited in queues, we need to provide the capacity to control  the number of running pod groups in queues with limited hardware resources.
**What you expected to happen**:
Add a parameter in Queue API:
```
type QueueSpec struct {
	Weight int32 `json:"weight,omitempty" protobuf:"bytes,1,opt,name=weight"`
         //  PodGroupNumber defines the max number of running podGroup in one queue, default to infinite.
        PodGroupNumber `json:"podgroupnumber,omitempty" protobuf:"bytes,1,opt,name=podgroupnumber"`
}
```
**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by hex108**
**What this PR does / why we need it**:
Add support for podGroup number limits for one queue

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #452

**Special notes for your reviewer**:
The PR add an admission webhook server. It is a start for other features, e.g. set default queue for each podgroup, set scheduler name based on namespace annotations. I'll add them after it is merged.

**Release note**:

```release-note
Add support for podGroup number limits for one queue
```

### CODE REVIEW

**Comment by hex108** [comment]
I have manually tested the feature. It works. I'll add doc and e2e tests in follow up PR.

The steps to make it work:

* Build the image

  `$ make images`

* Install kube-batch

  `$ helm install $GOPATH/src/github.com/kubernetes-sigs/kube-batch/deployment/kube-batch --namespace kube-system`

* Install the webhook server
  
   `$ helm install $GOPATH/src/github.com/kubernetes-sigs/kube-batch/deployment/webhook-server --namespace kube-system`

* Create following `Queue`

   ```
   $ cat queue.yaml
    apiVersion: scheduling.incubator.k8s.io/v1alpha1
    kind: Queue
    metadata:
         name: test-limit
    spec:
        weight: 2
        maxPodGroupNumber: 2
    $ kubectl create -f ./queue.yaml
   ```
   It will create a queue that limits the number of `PodGroup` in it to 2.

* Try to create 3 `PodGroup`s in the above `Queue`

   ```
   $ cat podgroup1.yaml
   apiVersion: scheduling.incubator.k8s.io/v1alpha1
   kind: PodGroup
   metadata:
        name: pg1
   spec:
       queue: test-limit
       minMember: 3
    $ kubect create -f ./podgroup1.yaml
    podgroup.scheduling.incubator.k8s.io/pg1 created
   ```

  ```
   $ cat podgroup2.yaml
   apiVersion: scheduling.incubator.k8s.io/v1alpha1
   kind: PodGroup
   metadata:
        name: pg2
   spec:
       queue: test-limit
       minMember: 3
    $ kubect create -f ./podgroup2.yaml
    podgroup.scheduling.incubator.k8s.io/pg2 created
   ```
   
    ```
   $ cat podgroup3.yaml
   apiVersion: scheduling.incubator.k8s.io/v1alpha1
   kind: PodGroup
   metadata:
        name: pg3
   spec:
       queue: test-limit
       minMember: 3
    $ kubect create -f ./podgroup3.yaml
    Error from server: error when creating "./podgroup3.yaml": admission webhook "webhook-server.kube-system.svc" denied the request: exceed MaxPodGroupNumber
   ```

   As noticed, we created `PodGroup` pg1 and pg2 successfully, but failed to create pg3. The feaure works!

**Comment by k82cn** [comment]
We may handle this in a similar way to Quota which need have Queue controller firstly; as we also support normal k8s object in kube-batch which belongs to default queue.

**Comment by hex108** [comment]
> We may handle this in a similar way to Quota which need have Queue controller firstly; as we also support normal k8s object in kube-batch which belongs to default queue.

What is Queue controller's responsibility?

File: deployment/kube-batch/templates/scheduling_v1alpha1_queue.yaml [hunk 1]
```diff
@@ -24,6 +24,9 @@ spec:
             weight:
               format: int32
               type: integer
+            maxPodGroupNumber:
```
Comment (k82cn) [inline deployment/kube-batch/templates/scheduling_v1alpha1_queue.yaml]: s/maxPodGroupNumber/capability/

File: deployment/webhook-server/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1alpha1
```
Comment (k82cn) [inline deployment/webhook-server/Chart.yaml]: it's better to merge this chart into the default one.

File: pkg/apis/scheduling/v1alpha1/types.go [hunk 3]
```diff
@@ -182,6 +182,8 @@ type Queue struct {
 // QueueSpec represents the template of Queue.
 type QueueSpec struct {
 	Weight int32 `json:"weight,omitempty" protobuf:"bytes,1,opt,name=weight"`
+	// MaxPodGroupNumber defines the max number of running podGroup in one queue, default to infinite.
+	MaxPodGroupNumber int32 `json:"maxPodGroupNumber,omitempty" protobuf:"bytes,1,opt,name=maxPodGroupNumber"`
```
Comment (k82cn) [inline pkg/apis/scheduling/v1alpha1/types.go]: `Capability` is better :)

File: pkg/webhook/validating/podgroup.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		return false, nil
+	}
+	// TODO: If the number of PodGroups is large, we should do some optimization, e.g.using informer and cache
+	// to implement it.
+	max := queue.Spec.MaxPodGroupNumber - 1
+	podGroups, err := pgv.admitterContext.PodGroupLister.List(labels.Everything())
+	if err != nil {
+		return true, err
+	}
+	for _, group := range podGroups {
```
Comment (k82cn) [inline pkg/webhook/validating/podgroup.go]: As we discussed offline, it's better to add a status to Queue and check it here, similar to Quota.

File: pkg/apis/scheduling/v1alpha1/types.go (hunk 3 above)
Comment (hex108) [reply pkg/apis/scheduling/v1alpha1/types.go]: Is `Capability` just for limiting pod group number? Then its type is `int32`. Or  will `Capability` contain other resources, e.g. `CPU`, `Mem`?

Comment (k82cn) [reply pkg/apis/scheduling/v1alpha1/types.go]: also contain other resources for extensibility, e.g. v1.Resources.

File: deployment/kube-batch/templates/scheduling_v1alpha1_queue.yaml (hunk 1 above)
Comment (hex108) [reply deployment/kube-batch/templates/scheduling_v1alpha1_queue.yaml]: addressed

File: deployment/webhook-server/Chart.yaml (hunk 2 above)
Comment (hex108) [reply deployment/webhook-server/Chart.yaml]: addressed

**Comment by hex108** [comment]
@k82cn PTAL

**Comment by hex108** [comment]
Close it since it is not needed.

/close

**Pull request closed without merging**